grobid-core/src/main/java/org/grobid/core/engines/FullTextParser.java

```diff
@@ -300,8 +348,72 @@ else if (config.getConsolidateCitations() == 2)
+            return null;
+
+        SortedSet<DocumentPiece> documentParts = new TreeSet<DocumentPiece>();
+        // identify continuous sequence of layout tokens in the abstract
+        int posStartPiece = -1;
+        int currentOffset = -1;
+        int startBlockPtr = -1;
+        LayoutToken previousToken = null;
+        for(LayoutToken token : tokens) {
+            if (currentOffset == -1) {
+                currentOffset = token.getOffset();
+                posStartPiece = getDocIndexToken(doc, token);
+                startBlockPtr = token.getBlockPtr();
+            } else {
+                if (token.getOffset() != currentOffset + previousToken.getText().length()) {
+                    // new DocumentPiece to be added 
+                    DocumentPointer dp1 = new DocumentPointer(doc, startBlockPtr, posStartPiece);
+                    DocumentPointer dp2 = new DocumentPointer(doc, 
+                        previousToken.getBlockPtr(), 
+                        getDocIndexToken(doc, previousToken));
+                    DocumentPiece piece = new DocumentPiece(dp1, dp2);
+                    documentParts.add(piece);
+
+                    // set index for the next DocumentPiece
+                    currentOffset = token.getOffset();
+                    posStartPiece = getDocIndexToken(doc, token);
+                    startBlockPtr = token.getBlockPtr();
+                } 
+            }
+            currentOffset = token.getOffset();
+            previousToken = token;
+        }
+        // we still need to add the last document piece
+        // conditional below should always be true because abstract is not null if we reach this part, but paranoia is good when programming 
+        if (posStartPiece != -1) {
+            DocumentPointer dp1 = new DocumentPointer(doc, startBlockPtr, posStartPiece);
+            DocumentPointer dp2 = new DocumentPointer(doc, 
+                previousToken.getBlockPtr(), 
+                getDocIndexToken(doc, previousToken));
+            DocumentPiece piece = new DocumentPiece(dp1, dp2);
+            documentParts.add(piece);
+        }
+
+        Pair<String, LayoutTokenization> featSeg = getBodyTextFeatured(doc, documentParts);
+        String res = null;
+        List<LayoutToken> tokenizations = null;
+        if (featSeg != null) {
+            // if featSeg is null, it usually means that no body segment is found in the
+            // document segmentation
+            String processedText = featSeg.getLeft();
+            tokenizations = featSeg.getRight().getTokenization();
+            // here, tokenizations should be identical to the provided tokens
+            if (isNotEmpty(trim(processedText))) 
+                res = label(processedText);
+        } else
+            return null;
+
+        return Pair.of(res, tokenizations);
+    }
+
+
     public Pair<String, List<LayoutToken>> processShort(List<LayoutToken> tokens, Document doc) {
```

Comment:
I'm not sure, but in this method `processShort` there is `if (currentPos+1 == tokenPos) {`. 

I have a doubt, but looking at the result of 
```GrobidAnalyzer.getInstance().tokenizeWithLayoutToken("This is a sentence")``` 

I get that `GrobidAnalyzer.getInstance().tokenizeWithLayoutToken("This is a sentence").get(1).getOffset()` is `4`, (`0 + lenght(This)`) so I'm not sure the `+1` is need there... two continuous tokens should be just `previousOffset + previousLenght == currentOffset`. 

Does it make sense? 

As a side note, I've discovered it when trying to write a unit test on the code that identifies the DocumentPieces. If you want I can push it (together with having moved the specific code that calculate the `DocumenPieces` on a separate method).

Reply:
I have somehow ended up again on this problem, and usually `currentPos + textLenght == nextToken.getOffset()` 🙂